File: src/main/java/org/athenian/java/OneWorkerMultiNotifier.java
package org.athenian.java;

import java.util.Random;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class OneWorkerMultiNotifier {

    public static void main(String[] args) {

      final ExecutorService executor = Executors.newCachedThreadPool();
      final Object monitor = new Object();
      final Random random = new Random();

      executor.submit(() -> {
        while (true) {
          System.out.println("Worker thread waiting...");
          try {
            synchronized (monitor) {
              monitor.wait();
            }
          } catch (InterruptedException e) {
            e.printStackTrace();
          }

          System.out.println("Worker thread done waiting on monitor");
        }
      });

      for (int j = 0; j < 3; j++) {
        int i = j;
        executor.submit(() -> {
          while (true) {
            long sleepTime = random.nextInt(5);
            System.out.printf("Notifier thread(%d) sleeping %d secs...%n", i, sleepTime);
            Utils.sleepSecs(sleepTime);

            System.out.printf("Notifier thread(%d) calling notify()%n", i);
            synchronized (monitor) {
              monitor.notify();
            }
          }
        });
      }
    }
}

File: src/main/kotlin/org/athenian/kotlin/OneWorkerMultiNotifier.kt
package org.athenian.kotlin

import java.util.concurrent.Executors
import kotlin.random.Random

fun main() {

    val executor = Executors.newCachedThreadPool()
    val monitor = Object()

    executor.submit {
        while (true) {
            println("Worker thread waiting...")
            try {
                synchronized(monitor) {
                    monitor.wait()
                }
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }

            println("Worker thread done waiting on monitor")
        }
    }

    (0..3).forEach { i ->
        executor.submit {
            while (true) {
                val sleepTime = Random.nextInt(5).toLong()
                println("Notifier thread($i) sleeping $sleepTime secs...")
                sleepSecs(sleepTime)

                println("Notifier thread($i) calling notify()")
                synchronized(monitor) {
                    monitor.notify()
                }
            }
        }
    }
}
